File: Trees/112: Path Sum.js
// https://leetcode.com/problems/path-sum/description/
// Difficulty: Easy
// tags: dfs

// Problem

/*
Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.

A leaf is a node with no children.
*/

// Solution, O(n) time and O(n) (height) space for the callstack

/*
DFS throughout the tree, if we have the target sum and are at a leaf, return true. Otherwise return the results of either of the children.
*/

var hasPathSum = function (root, targetSum) {
  function dfs(node, currentSum) {
    if (!node) {
      return false;
    }

    currentSum += node.val;

    if (currentSum === targetSum && !node.left && !node.right) {
      return true;
    }

    return dfs(node.left, currentSum) || dfs(node.right, currentSum);
  }

  return dfs(root, 0);
};

File: problems/Leetcode/112: Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        return ((self.hasPathSum(root.left, targetSum - root.val) or 
        self.hasPathSum(root.right, targetSum - root.val))
        if root and (root.left or root.right) else targetSum == root.val if root else False)